hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/RateLimiter.java

```diff
@@ -135,18 +135,31 @@ protected synchronized long getTimeUnitInMillis() {
 
   /**
    * Is there at least one resource available to allow execution?
-   * @return true if there is at least one resource available, otherwise false
+   * @return the waitInterval to backoff, or 0 if execution is allowed
    */
-  public boolean canExecute() {
+  public long canExecute() {
     return canExecute(1);
   }
 
+  /**
+   * Are there enough available resources to allow execution?
+   * @param amount the number of required resources, a non-negative number
+   * @return the waitInterval to backoff, or 0 if execution is allowed
+   */
+  public synchronized long canExecute(final long amount) {
+    assert amount >= 0;
+    if (!isAvailable(amount)) {
+      return waitInterval(amount);
+    }
+    return 0;
+  }
+
   /**
    * Are there enough available resources to allow execution?
    * @param amount the number of required resources, a non-negative number
    * @return true if there are enough available resources, otherwise false
    */
-  public synchronized boolean canExecute(final long amount) {
+  private synchronized boolean isAvailable(final long amount) {
```

Comment:
Since we already have synchronization in the caller above, do we still need synchronization here?

For me, I prefer we only add synchronization in the public methods, which could make it more clear on what we are protecting.